### PR TITLE
feat: Added Terraform config for an Elasticache Redis single instance cluster [staging env]

### PR DESCRIPTION
**Pull request by jimleroyer**
# Summary | Résumé

[Following this ticket](https://trello.com/c/Hb6ijLTD/202-use-aws-elasticache-as-our-redis-instance), we want to switch from a self managed Redis instance to an AWS Elasticache Redis instance.

# Test instructions | Instructions pour tester la modification

Once regular terraform validation have passed and changes have been deployed, log into the AWS staging console to inspect the expected changes.

# Unresolved questions / Out of scope | Questions non résolues ou hors sujet

To monitor the current state of the ElastiCache node:

* https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/ECEvents.html 

# Reviewer checklist | Liste de vérification du réviseur

This is a suggested checklist of questions reviewers might ask during their review | Voici une suggestion de liste de vérification comprenant des questions que les réviseurs pourraient poser pendant leur examen :

- [ ] Does this meet a user need? | Est-ce que ça répond à un besoin utilisateur?
- [ ] Is it accessible? | Est-ce que c’est accessible?
- [ ] Is it translated between both offical languages? | Est-ce dans les deux langues officielles?
- [ ] Is the code maintainable? | Est-ce que le code peut être maintenu?
- [ ] Have you tested it? | L’avez-vous testé?
- [ ] Are there automated tests? | Y a-t-il des tests automatisés?
- [ ] Does this cause automated test coverage to drop? | Est-ce que ça entraîne une baisse de la quantité de code couvert par les tests automatisés?
- [ ] Does this break existing functionality? | Est-ce que ça brise une fonctionnalité existante?
- [ ] Should this be split into smaller PRs to decrease change risk? | Est-ce que ça devrait être divisé en de plus petites demandes de tirage (« pull requests ») afin de réduire le risque lié aux modifications?
- [ ] Does this change the privacy policy? | Est-ce que ça entraîne une modification de la politique de confidentialité?
- [ ] Does this introduce any security concerns? | Est-ce que ça introduit des préoccupations liées à la sécurité?
- [ ] Does this significantly alter performance? | Est-ce que ça modifie de façon importante la performance?
- [ ] What is the risk level of using added dependencies? | Quel est le degré de risque d’utiliser des dépendances ajoutées?
- [ ] Should any documentation be updated as a result of this? (i.e. README setup, etc.) | Faudra-t-il mettre à jour la documentation à la suite de ce changement (fichier README, etc.)?

### CODE REVIEW

**Comment by maxneuvians** [comment]
Just as an aside I have found that you need to bind resource dependent security groups into the main EKS security group, otherwise the EKS cluster has no way to communicate with the resource. We had this issue with RDS https://github.com/cds-snc/notification-terraform/blob/main/aws/rds/rds.tf#L48-L50 and the ALB https://github.com/cds-snc/notification-terraform/blob/main/aws/eks/alb.tf#L11. This means that the EKS security group needs to exist before you can provision the redis cluster. This means you may need to take it out of `common` and write a new elasticache module like the `rds` one.

**Comment by jimleroyer** [comment]
Thank for the heads up omniscient @maxneuvians . I will put this into a different module then. 👏

**Comment by maxneuvians** [comment]
As mentioned ideally elasticache would have its own security group and you would just allow inbound/outbound traffic to the EKS one. Somehow it never really worked, might just need to dig a little deeper.

File: aws/elasticache/elasticache.tf [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    # is deprecated by AWS. Version 6 also has lot more advantages, if we're
+    # willing to do the upgrade in this dev iteration. [3]
+    #
+    # [1] https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/supported-engine-versions.html#redis-version-3-2-10
+    # [2] https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/supported-engine-versions.html#redis-version-4-0-10
+    # [3] https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/supported-engine-versions.html
+    engine_version       = "4.0.10"
+    # REVIEW: RDS maintenance is on Wed, this one Thursday,
+    # or we prefer the same time window for both?
+    maintenance_window   = "thu:04:00-thu:05:00"
```
Comment (AntoineAugusti) [inline aws/elasticache/elasticache.tf:31]: (You can put a comment on your PR next time if you're not sure of something in the code 👍 )

Picking a different day for the maintenance window seems like a good idea! Is it possible to subscribe to some maintenance events to get a heads up before?

It's possible for RDS https://github.com/cds-snc/notification-terraform/blob/8a859b0028597f6a90a115edaefc6d85a54536a6/aws/rds/rds.tf#L65-L79

File: aws/elasticache/elasticache.tf [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    cluster_id           = "notification-cluster-cache"
+    engine               = "redis"
+    # The nearest versions of Redis supported by AWS next to the one we
+    # use (i.e. v3.5.3) would be 3.2.10 or 4.0.10. The former does not
+    # support in-transit and at rest encryption whereas the latter does.
+    # Hence I picked the higher major version for now. [1][2]
+    #
+    # One also have to be useful when defining the version to check if it
+    # is deprecated by AWS. Version 6 also has lot more advantages, if we're
+    # willing to do the upgrade in this dev iteration. [3]
```
Comment (AntoineAugusti) [inline aws/elasticache/elasticache.tf:23]: We don't use strange features of Redis so I think we're fine with changing versions now. Can you pick the latest available version on AWS?

File: aws/elasticache/elasticache.tf [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+resource "aws_elasticache_subnet_group" "notification-canada-ca-cache-subnet" {
+    name       = "notification-canada-ca-${var.env}-cache-subnet"
+    subnet_ids = var.vpc_private_subnets
+}
+
+resource "aws_elasticache_cluster" "notification-cluster-cache" {
+    # Default is false with this param, it looks counter-intuitive because
+    # applied changes would only happen during maintenance window if false.
+    apply_immediately    = true
+    cluster_id           = "notification-cluster-cache"
```
Comment (AntoineAugusti) [inline aws/elasticache/elasticache.tf:14]:
```suggestion
    cluster_id           = "notification-canada-ca-${var.env}-cluster-cache"
```
I think we put the env in all our names

File: env/production/common/terragrunt.hcl [hunk 4]
```diff
@@ -9,5 +9,5 @@ include {
 }
 
 inputs = {
-  sns_monthly_spend_limit = 10000
+  sns_monthly_spend_limit  = 10000
```
Comment (AntoineAugusti) [inline env/production/common/terragrunt.hcl:12]: Can you run a `make fmt`? I think this space should not be here

File: env/staging/elasticache/terragrunt.hcl [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+include {
+  path = find_in_parent_folders()
+}
+
+inputs = {
+  eks_cluster_securitygroup = dependency.eks.outputs.eks-cluster-securitygroup
+  elasticache_node_type     = "cache.t3.micro"
+  sns_alert_warning_arn     = dependency.common.outputs.sns_alert_warning_arn
```
Comment (AntoineAugusti) [inline env/staging/elasticache/terragrunt.hcl:39]: I don't think this variable is used at the moment

File: env/production/elasticache/terragrunt.hcl [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+include {
+  path = find_in_parent_folders()
+}
+
+inputs = {
+  eks_cluster_securitygroup = dependency.eks.outputs.eks-cluster-securitygroup
+  elasticache_node_type     = "cache.t3.micro"
+  sns_alert_warning_arn     = dependency.common.outputs.sns_alert_warning_arn
```
Comment (AntoineAugusti) [inline env/production/elasticache/terragrunt.hcl:24]: I don't think this is used

File: aws/elasticache/elasticache.tf [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    #
+    # [1] https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/supported-engine-versions.html#redis-version-3-2-10
+    # [2] https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/supported-engine-versions.html#redis-version-4-0-10
+    # [3] https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/supported-engine-versions.html
+    engine_version       = "4.0.10"
+    # REVIEW: RDS maintenance is on Wed, this one Thursday,
+    # or we prefer the same time window for both?
+    maintenance_window   = "thu:04:00-thu:05:00"
+    node_type            = var.elasticache_node_type
+    num_cache_nodes      = 1
```
Comment (AntoineAugusti) [inline aws/elasticache/elasticache.tf:33]: Can you put this in a variable?

**Review by AntoineAugusti — CHANGES_REQUESTED** [review]
Good work 👏 

A few things to update before merging the PR

File: aws/elasticache/elasticache.tf (hunk 1 above)
Comment (jimleroyer) [reply aws/elasticache/elasticache.tf:31]: A search for the equivalent feature was not successful from my side, even though it does exist for redshift and other tech. I am skeptical it really is not there in Terraform as support [exists in AWS](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/ElastiCacheSNS.html) and it seems it is not mentioned in the [Terraform issue tracker](https://github.com/hashicorp/terraform/issues?q=is%3Aissue+%22elasticache%22+event+). I will probably open an issue myself and see what is the reply.

Would there be any other way so manually subscribe to the events and publish these on sns?

File: aws/elasticache/elasticache.tf [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  apply_immediately = true
+  cluster_id        = "notification-canada-ca-${var.env}-cluster-cache"
+  engine            = "redis"
+  # AWS automatically supports the Redis minor version management since version 6.
+  # https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/supported-engine-versions.html#redis-version-6.x
+  engine_version       = "6.x"
+  maintenance_window   = "thu:04:00-thu:05:00"
+  node_type            = var.elasticache_node_type
+  num_cache_nodes      = var.elasticache_node_count
+  parameter_group_name = "default.redis4.0"
```
Comment (AntoineAugusti) [inline aws/elasticache/elasticache.tf:22]:
```suggestion
  parameter_group_name = "default.redis6.x"
```

File: env/staging/elasticache/terragrunt.hcl [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  # Configure mock outputs for the `validate` command that are returned when there are no outputs available (e.g the
+  # module hasn't been applied yet.
+  mock_outputs_allowed_terraform_commands = ["validate"]
+  mock_outputs = {
+    vpc_private_subnets = [
+      "subnet-001e585d12cce4d1e",
+      "subnet-08de34a9e1a7458dc",
+      "subnet-0af8b8402f1d605ff",
+    ]
+    sns_alert_warning_arn = ""
```
Comment (AntoineAugusti) [inline env/staging/elasticache/terragrunt.hcl:17]: This is not used

Comment (jimleroyer) [reply env/staging/elasticache/terragrunt.hcl:17]: Just removed..

**Review by AntoineAugusti — APPROVED** [review]
This can be shipped in staging 👍 

Before merging, can you remove changes you made to these files:
- .github/workflows/merge_to_main_production.yml
- env/production/elasticache/terragrunt.hcl

These files are fine but we can't release them now. If we merge as is, it'll try to deploy to production with the current `INFRASTRUCTURE_VERSION` tag. This tag does not include the new Elasticache changes so we'll need to:
- merge this PR, this will create a new tag
- if this is okay, release to production by bumping the tag and making the changes you just proposed to these files

**Review by AntoineAugusti — APPROVED** [review]
🚢

**Pull request merged**